### DISCUSSION TITLE
Struggling with xml put, no value

### DISCUSSION THREAD

**Discussion by yorickdowne** [TomWright/dasel#220] [Q&A] [answered]
Given an NLog sample.xml like this:

```
<nlog>
    <rules>
        <logger name="JsonWebAPI.Microsoft.Extensions.Diagnostics.HealthChecks.DefaultHealthCheckService" maxLevel="Error" final="true" />
        <logger name="JsonWebAPI*" minlevel="Error" writeTo="file-async" />
        <logger name="JsonWebAPI*" minlevel="Error" writeTo="auto-colored-console-async" final="true" />
        <logger name="JsonWebAPI*" final="true" />
        <logger name="*" minlevel="Off" writeTo="seq" />
        <logger name="*" minlevel="Info" writeTo="file-async" />
        <logger name="*" minlevel="Info" writeTo="auto-colored-console-async" />
    </rules>
</nlog>
```

I'd like to add more `<logger>` entries but am failing. The docs get me only so far. I can successfully `dasel -f sample.xml -p xml '.nlog.rules'` but if I query say `-name` it fails, because these nodes have no value:

```
:~$ dasel -f sample.xml -p xml '.nlog.rules.logger[*].-name'
Error: could not query node: could not find value: no value found for selector: .logger[*]: map[logger:[map[-final:true -maxLevel:Error -name:JsonWebAPI.Microsoft.Extensions.Diagnostics.HealthChecks.DefaultHealthCheckService] map[-minlevel:Error -name:JsonWebAPI* -writeTo:file-async] map[-final:true -minlevel:Error -name:JsonWebAPI* -writeTo:auto-colored-console-async] map[-final:true -name:JsonWebAPI*] map[-minlevel:Off -name:* -writeTo:seq] map[-minlevel:Info -name:* -writeTo:file-async] map[-minlevel:Info -name:* -writeTo:auto-colored-console-async]]]
```

Ultimately I want to `put` something here, but am struggling with the correct syntax. For example, to put `<logger name="JsonRpc.*" minlevel="Error" writeTo="file-async"/>` I tried the following. I know the syntax is entirely wrong, I just cannot figure out what it should be.

`dasel put object -t string -f sample.xml -o out.xml -p xml '.nlog.rules.logger[]' 'logger.-name="JsonRpc.*" logger.-minlevel="Error" loger.-writeTo="file-async" ""'`

Result:
```
<nlog>
  <rules>
      <logger final="true" maxLevel="Error" name="JsonWebAPI.Microsoft.Extensions.Diagnostics.HealthChecks.DefaultHealthCheckService"/>
    <logger minlevel="Error" name="JsonWebAPI*" writeTo="file-async"/>
    <logger final="true" minlevel="Error" name="JsonWebAPI*" writeTo="auto-colored-console-async"/>
    <logger final="true" name="JsonWebAPI*"/>
    <logger minlevel="Off" name="*" writeTo="seq"/>
    <logger minlevel="Info" name="*" writeTo="file-async"/>
    <logger minlevel="Info" name="*" writeTo="auto-colored-console-async"/>
    <logger[]>
      <logger.-name>"JsonRpc.*" logger.-minlevel="Error" loger.-writeTo="file-async" ""</logger.-name>
    </logger[]>
  </rules>
</nlog>
```

Ideally I want no value and a new `logger` entry with attributes.

**Comment by TomWright** [comment, marked as answer]
Hey, thanks for the question.

I've had a play around and the issue here is with the XML parser providing somewhat of a ghost root node.

A simple workaround that works is the following:
```bash
dasel put document -f sample.xml -d json -o out.xml 'nlog.rules.logger.[]' '{"-name":"JsonRpc.*","-minlevel":"Error","-writeTo":"file-async"}'
```

Example:

The original file:
```bash
$ cat sample.xml
<nlog>
    <rules>
        <logger name="JsonWebAPI.Microsoft.Extensions.Diagnostics.HealthChecks.DefaultHealthCheckService" maxLevel="Error" final="true" />
        <logger name="JsonWebAPI*" minlevel="Error" writeTo="file-async" />
        <logger name="JsonWebAPI*" minlevel="Error" writeTo="auto-colored-console-async" final="true" />
        <logger name="JsonWebAPI*" final="true" />
        <logger name="*" minlevel="Off" writeTo="seq" />
        <logger name="*" minlevel="Info" writeTo="file-async" />
        <logger name="*" minlevel="Info" writeTo="auto-colored-console-async" />
    </rules>
</nlog>
```

Using dasel:
```bash
$ dasel put document -f sample.xml -d json -o out.xml 'nlog.rules.logger.[]' '{"-name":"JsonRpc.*","-minlevel":"Error","-writeTo":"file-async"}'
```

The output:
```bash
$ cat out.xml
<nlog>
  <rules>
      <logger final="true" maxLevel="Error" name="JsonWebAPI.Microsoft.Extensions.Diagnostics.HealthChecks.DefaultHealthCheckService"/>
    <logger minlevel="Error" name="JsonWebAPI*" writeTo="file-async"/>
    <logger final="true" minlevel="Error" name="JsonWebAPI*" writeTo="auto-colored-console-async"/>
    <logger final="true" name="JsonWebAPI*"/>
    <logger minlevel="Off" name="*" writeTo="seq"/>
    <logger minlevel="Info" name="*" writeTo="file-async"/>
    <logger minlevel="Info" name="*" writeTo="auto-colored-console-async"/>
    <logger minlevel="Error" name="JsonRpc.*" writeTo="file-async"/>
  </rules>
</nlog>
```

Honestly I'm not sure why the spacing on the first item is off though, I'm happy to keep this issue open while I do some digging but this should unblock you short term.

**Reply by TomWright** [reply]
I also acknowledge that the XML support here is a little iffy. It's hard to deal with using the existing parser

**Reply by yorickdowne** [reply]
I can see that. `dasel -f sample.xml .nlog.rules` works, but `dasel -f sample.xml .nlog.rules.logger[]` fails with `Error: could not query node: could not find value: no value found for selector:`

**Reply by TomWright** [reply]
There is an issue with your selector. Note that `logger` and `[]` must be separated by a `.`.

Also, the `[]` selector is used to select the next unused index which when selecting data is invalid. If you want to view all loggers you can use `.nlog.rules.logger.[*]`. This should be combined with the `-m` flag to return multiple results.

https://dasel.tomwright.me/s/a394b8ac-a57c-4ffc-a10a-79a6472e9832

If this doesn't help please let me know what you are trying to get out and I can provide a solution.

**Reply by yorickdowne** [reply]
Thanks again! That does it. There's a depth to dasel that I am still struggling with. I appreciate you taking time to explain it to me. I do see the dot in the docs, now. 

And I know where I got confused: https://daseldocs.tomwright.me/examples/xml uses `.data.users[0].-primary` without the dot and I've been following that pattern ever since.

**Reply by TomWright** [reply]
Ah that is actually a typo, I'll fix it now

**Comment by yorickdowne** [comment]
Thanks a ton for taking the time to dig into  that! I wouldn't have figured that out in an age.
